servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,619 @@
+		BARRIER_ACCESS_COLOR_ATTACHMENT_WRITE_BIT = (1 << 8),
+		BARRIER_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT = (1 << 9),
+		BARRIER_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT = (1 << 10),
+		BARRIER_ACCESS_TRANSFER_READ_BIT = (1 << 11),
+		BARRIER_ACCESS_TRANSFER_WRITE_BIT = (1 << 12),
+		BARRIER_ACCESS_HOST_READ_BIT = (1 << 13),
+		BARRIER_ACCESS_HOST_WRITE_BIT = (1 << 14),
+		BARRIER_ACCESS_MEMORY_READ_BIT = (1 << 15),
+		BARRIER_ACCESS_MEMORY_WRITE_BIT = (1 << 16),
+		BARRIER_ACCESS_FRAGMENT_SHADING_RATE_ATTACHMENT_READ_BIT = (1 << 23),
+	};
+
+	struct MemoryBarrier {
+		BitField<BarrierAccessBits> src_access;
+		BitField<BarrierAccessBits> dst_access;
+	};
+
+	struct BufferBarrier {
+		BufferID buffer = 0;
+		BitField<BarrierAccessBits> src_access;
+		BitField<BarrierAccessBits> dst_access;
+		uint64_t offset = 0;
+		uint64_t size = 0;
+	};
+
+	struct TextureBarrier {
+		TextureID texture = 0;
+		BitField<BarrierAccessBits> src_access;
+		BitField<BarrierAccessBits> dst_access;
+		TextureLayout prev_layout = TEXTURE_LAYOUT_UNDEFINED;
+		TextureLayout next_layout = TEXTURE_LAYOUT_UNDEFINED;
+		TextureSubresourceRange subresources;
+	};
+
+	virtual void command_pipeline_barrier(
+			CommandBufferID p_cmd_buffer,
+			BitField<PipelineStageBits> p_src_stages,
+			BitField<PipelineStageBits> p_dst_stages,
+			VectorView<MemoryBarrier> p_memory_barriers,
+			VectorView<BufferBarrier> p_buffer_barriers,
+			VectorView<TextureBarrier> p_texture_barriers) = 0;
+
+	/*************************/
+	/**** COMMAND BUFFERS ****/
+	/*************************/
+
+	// ----- POOL -----
+
+	virtual CommandPoolID command_pool_create() = 0;
+	virtual void command_pool_free(CommandPoolID p_cmd_pool) = 0;
+
+	// ----- BUFFER -----
+
+	enum CommandBufferType {
+		COMMAND_BUFFER_TYPE_PRIMARY,
+		COMMAND_BUFFER_TYPE_SECONDARY,
+	};
+
+	virtual CommandBufferID command_buffer_create(CommandBufferType p_cmd_buffer_type, CommandPoolID p_cmd_pool) = 0;
+	virtual bool command_buffer_begin(CommandBufferID p_cmd_buffer) = 0;
+	virtual bool command_buffer_begin_secondary(CommandBufferID p_cmd_buffer, RenderPassID p_render_pass, uint32_t p_subpass, FramebufferID p_framebuffer) = 0;
+	virtual void command_buffer_end(CommandBufferID p_cmd_buffer) = 0;
+	virtual void command_buffer_execute_secondary(CommandBufferID p_cmd_buffer, VectorView<CommandBufferID> p_secondary_cmd_buffers) = 0;
+
+	/*********************/
+	/**** FRAMEBUFFER ****/
+	/*********************/
+
+	virtual FramebufferID framebuffer_create(RenderPassID p_render_pass, VectorView<TextureID> p_attachments, uint32_t p_width, uint32_t p_height) = 0;
+	virtual void framebuffer_free(FramebufferID p_framebuffer) = 0;
+
+	/****************/
+	/**** SHADER ****/
+	/****************/
+
+	virtual String shader_get_binary_cache_key() = 0;
+	virtual Vector<uint8_t> shader_compile_binary_from_spirv(VectorView<ShaderStageSPIRVData> p_spirv, const ShaderReflection &p_shader_refl, const String &p_shader_name) = 0;
```

Comment:
So far this is the responsibility of the backend, because it returns the format that is best preferred for caching in the driver for binary.

That said, when exporting we may want to ship caches precompiled, so these compilation things may need to be able from somewhere else, but at the moment it does not matter and I would leave as is. That's Pedro's problem when he does this export optimization.

Reply:
For the time being, yes, everything orbits around SPIR-V. That said, the part that is API-agnostic, which is the reflection into Godot `RenderingDevice` structures is performed by `RenderingDevice`. From that info, the driver has to put it into it's own format. The D3D12 backend does the translation to DXIL as part of that process. Honestly, other parts of the process of writing the Godot shader binary, like compressing the bytecode of the stages, etc. is very similar or identical between Vulkan and D3D12, and can probably be refactored into common and distinct parts... in a future iteration.